File: Python/Easy/1768_merge_strings_alternately.py
"""
1768. Merge Strings Alternately


Problem Statement:
You are given two strings word1 and word2. Merge the strings by adding letters
in alternating order, starting with word1. If a string is longer than the other,
append the additional letters onto the end of the merged string.

Return the merged string.

Constraints:
- 1 <= word1.length, word2.length <= 100
- word1 and word2 consist of lowercase English letters.

Examples:

Example 1:
    Input: word1 = "abc", word2 = "pqr"
    Output: "apbqcr"

Example 2:
    Input: word1 = "ab", word2 = "pqrs"
    Output: "apbqrs"

Example 3:
    Input: word1 = "abcd", word2 = "pq"
    Output: "apbqcd"
"""

from itertools import zip_longest


class Solution:
    def mergeAlternatelyMap(self, word1: str, word2: str) -> str:
        """
        Merges two strings alternately using map and lambda.

        Time Complexity: O(n1 + n2), where n1 is the length of word1, and n2
        is the length of word2. Space Complexity: O(n1 + n2), for storing the
        final merged string.
        """
        max_len = max(len(word1), len(word2))

        merged = ''.join(map(lambda i: (word1[i] if i < len(word1) else '') +
                                         (word2[i] if i < len(word2) else ''), range(max_len)))

        return merged

    def mergeAlternatelyWhile(self, word1: str, word2: str) -> str:
        """
        Merges two strings alternately using a while loop.

        Time Complexity: O(n1 + n2), where n1 is the length of word1, and n2 is
        the length of word2. Space Complexity:O(n1 + n2), for storing the final
        merged string.
        """
        merged = []
        i, j = 0, 0
        n1, n2 = len(word1), len(word2)

        while i < n1 or j < n2:
            if i < n1:
                merged.append(word1[i])
                i += 1
            if j < n2:
                merged.append(word2[j])
                j += 1

        return ''.join(merged)

    def mergeAlternately(self, word1: str, word2: str) -> str:
        """
        Merges two strings alternately using three explicit while loops.

        Uses a flag to alternate between words while both have characters
        remaining, then two separate loops to append any leftover characters.
        More explicit than the single-loop approach — preferred for interviews
        where you cannot rely on stdlib helpers.

        Time complexity:  O(n1 + n2)
        Space complexity: O(n1 + n2)
        """
        A, B = len(word1), len(word2)
        a, b = 0, 0
        s = []

        word = 1
        while a < A and b < B:
            if word == 1:
                s.append(word1[a])
                a += 1
                word = 2
            else:
                s.append(word2[b])
                b += 1
                word = 1

        while a < A:
            s.append(word1[a])
            a += 1

        while b < B:
            s.append(word2[b])
            b += 1

        return ''.join(s)

    def mergeAlternatelyZipLongest(self, word1: str, word2: str) -> str:
        """
        Merges two strings alternately using zip_longest from itertools.

        zip_longest pairs characters from both strings, filling the shorter
        one with an empty string via fillvalue=''. Each pair is concatenated
        and joined into the final result. The most Pythonic approach — but
        requires an import, so not suitable for environments that disallow
        standard library use.

        Time complexity:  O(n1 + n2)
        Space complexity: O(n1 + n2)
        """
        return ''.join(a + b for a, b in zip_longest(word1, word2, fillvalue=''))


if __name__ == "__main__":
    # Test cases
    test_cases = [
        ("abc", "pqr", "apbqcr"),
        ("ab", "pqrs", "apbqrs"),
        ("abcd", "pq", "apbqcd")
    ]

    solution = Solution()

    for label, method in [
        ("mergeAlternately           ", solution.mergeAlternately),
        ("mergeAlternatelyWhile      ", solution.mergeAlternatelyWhile),
        ("mergeAlternatelyMap        ", solution.mergeAlternatelyMap),
        ("mergeAlternatelyZipLongest ", solution.mergeAlternatelyZipLongest),
    ]:
        for w1, w2, expected in test_cases:
            result = method(w1, w2)
            assert result == expected, \
                f"{label} failed on ({w1!r}, {w2!r}): expected {expected!r}, got {result!r}"
        print(f"  {label} passed")

    print("All test cases passed!")

File: Golang/Easy/1768_merge_strings_alternately.go
package main

import "fmt"

// Solution groups all approaches for 1768. Merge Strings Alternately.
//
// Given two strings word1 and word2, merge them by adding letters in
// alternating order starting with word1. If one string is longer, append the
// remaining letters at the end.
type Solution struct{}

// mergeAlternately merges word1 and word2 alternately using a two-pointer
// while loop over a pre-allocated byte slice.
//
// Uses []byte instead of a string builder or repeated concatenation to avoid
// allocating a new string on every append. The byte slice is pre-allocated
// with capacity n+m so no re-allocations occur during the loop.
// Casting to string(result) once at the end is the only final allocation.
//
// Time complexity:  O(n1 + n2)
// Space complexity: O(n1 + n2)
func (s Solution) mergeAlternately(word1, word2 string) string {
	n, m := len(word1), len(word2)
	result := make([]byte, 0, n+m)
	i, j := 0, 0

	for i < n || j < m {
		if i < n {
			result = append(result, word1[i])
			i++
		}
		if j < m {
			result = append(result, word2[j])
			j++
		}
	}

	return string(result)
}

// mergeAlternatelyMinLoop merges word1 and word2 alternately using a
// two-phase approach: one loop up to min(n, m) appending both characters at
// once, then a single append of the remaining tail.
//
// append(result, word1[i], word2[i]) adds two bytes in one call, avoiding a
// second append per iteration. word1[l:]... spreads the remaining bytes
// directly into the slice — no character-by-character loop for the tail.
// min() is a built-in as of Go 1.21.
//
// Time complexity:  O(n1 + n2)
// Space complexity: O(n1 + n2)
func (s Solution) mergeAlternatelyMinLoop(word1, word2 string) string {
	n, m := len(word1), len(word2)
	l := min(n, m)
	result := make([]byte, 0, n+m)

	for i := 0; i < l; i++ {
		result = append(result, word1[i], word2[i])
	}

	result = append(result, word1[l:]...)
	result = append(result, word2[l:]...)

	return string(result)
}

// mergeAlternatelyMap merges word1 and word2 alternately by iterating up to
// max(len(word1), len(word2)) and appending each character when in range.
//
// Time complexity:  O(n1 + n2)
// Space complexity: O(n1 + n2)
func (s Solution) mergeAlternatelyMap(word1, word2 string) string {
	return "" // TODO
}

func main() {
	testCases := []struct {
		word1    string
		word2    string
		expected string
	}{
		{word1: "abc", word2: "pqr", expected: "apbqcr"},
		{word1: "ab", word2: "pqrs", expected: "apbqrs"},
		{word1: "abcd", word2: "pq", expected: "apbqcd"},
	}

	sol := Solution{}

	for _, approach := range []struct {
		label string
		fn    func(string, string) string
	}{
		{"mergeAlternately      ", sol.mergeAlternately},
		{"mergeAlternatelyMinLoop", sol.mergeAlternatelyMinLoop},
		{"mergeAlternatelyMap   ", sol.mergeAlternatelyMap},
	} {
		fmt.Printf("--- %s ---\n", approach.label)
		for i, tc := range testCases {
			result := approach.fn(tc.word1, tc.word2)
			status := "PASS"
			if result != tc.expected {
				status = "FAIL"
			}
			fmt.Printf("  Example %d | Input: word1=%q word2=%q | Expected: %q | Result: %q | [%s]\n",
				i+1, tc.word1, tc.word2, tc.expected, result, status)
		}
	}
}
